include/exec/at_coroutine_exit.hpp

```diff
@@ -0,0 +1,256 @@
+ * limitations under the License.
+ */
+
+#include <exception>
+#include <type_traits>
+
+#include "stdexec/execution.hpp"
+#include "exec/task.hpp"
+
+namespace exec {
+namespace __at_coroutine_exit {
+struct __die_on_stop_t {
+  template <class _Receiver>
+    struct __receiver_id {
+      struct __t {
+        using __id = __receiver_id;
+        _Receiver __receiver_;
+
+        template <class... _Args>
+          friend void tag_invoke(stdexec::set_value_t, __t&& __self,
+                                _Args&&... __args) noexcept {
+            if constexpr (stdexec::__nothrow_callable<stdexec::set_value_t, _Receiver&&, _Args&&...>) {
+              stdexec::set_value((_Receiver &&) __self.__receiver_, (_Args &&) __args...);
+            } else {
+              try {
+                stdexec::set_value((_Receiver &&) __self.__receiver_, (_Args &&) __args...);
+              } catch (...) {
+                stdexec::set_error((_Receiver &&) __self.__receiver_, std::current_exception());
+              }
+            }
+          }
+
+        template <class _Error>
+          friend void tag_invoke(stdexec::set_error_t, __t&& __self,
+                                _Error&& __error) noexcept {
+            stdexec::set_error((_Receiver &&) __self.__receiver_, (_Error &&) __error);
+          }
+
+        [[noreturn]] friend void tag_invoke(stdexec::set_stopped_t, __t&&) noexcept {
+          std::terminate();
+        }
+
+        friend auto tag_invoke(stdexec::get_env_t, const __t& __self) noexcept {
+          return stdexec::get_env(__self.__receiver_);
+        }
+      };
+    };
+  template <class _R> using __receiver = stdexec::__t<__receiver_id<_R>>;
+
+  template <class _Sig> struct __return_type {
+    using type = _Sig;
+  };
+
+  template <class _R, class... _Args> struct __return_type<_R(_Args...)> {
+    using type = _R;
+  };
+
+  template <class _Sig>
+  using __return_type_t = typename __return_type<_Sig>::type;
+
+  template <class _Tag, class _With = stdexec::__> struct __replace_tag {
+    template <class _Arg>
+    using __f = stdexec::__if_c<std::is_same_v<__return_type_t<_Arg>, _Tag>, _With, _Arg>;
+  };
+
+  template <class _Tag>
+    struct __remove_signatures_with_tag {
+      template <class... _Args>
+      using __f =
+          stdexec::__mapply<
+              stdexec::__remove<stdexec::__, stdexec::__q<stdexec::completion_signatures>>,
+              stdexec::__minvoke<stdexec::__transform<__replace_tag<_Tag, stdexec::__>>, _Args...>>;
+    };
+
+  template <class _Sender>
+    struct __sender_id {
+      using __old_sigs = stdexec::__completion_signatures_of_t<_Sender, stdexec::env_of_t<_Sender>>;
+      using __completion_signatures_t =
+          stdexec::__mapply<__remove_signatures_with_tag<stdexec::set_stopped_t>, __old_sigs>;
+      class __t {
+      public:
+        using __id = __sender_id;
+
+        using completion_signatures = __completion_signatures_t;
```

Comment:
This part isn't right, I don't think. You should be computing the completion signatures using the receiver's environment by defining a `get_completion_signatures_t` tag_invoke overload.